### DISCUSSION TITLE
Ability to combine workflows or set global steps out of workflow

### DISCUSSION THREAD

**Discussion by Reedyuk** [codemagic-ci-cd/codemagic-docs#1819] [Ideas]
### Use case

```Text
I have several workflows, test pull request, run integration tests, release build.

They share several steps, so currently i have to define them in the first workflow to be able to reference them in another workflow.

My use case, I have a set of steps that run integration tests e.g. run emulator, run integration test command. These live in the integration tests workflow, but get used by *runIntegration test reference in other workflows.

Problem: If i decide to create another workflow in the codemagic.yaml file and say i put this new workflow at the top of file, it sucks because i have to move my reusable steps in the first workflow.
```

### Idea

```Text
2 possible solutions.

1. Ability to link another workflow, just like we do with script steps, where it will inehrit just the steps in the workflow. So then in my example, i could just call *runIntegrationTestsWorkflow like i would usually do in steps.

2. Ability to define a set of 'global' steps at the top of a workflow, then reference them in workflows like we normally do. This would mean then most workflows would look less busy and just reference the steps.
```

**Comment by mikhail-tokarev** [comment]
thanks for trying Codemagic and making all this proposals 🙌 would you be interesting to have a call to learn your case better? usually 20-30 minutes is enough.

re your question: this is completely on YAML side and for instance internally we have `definitions` section to include everything we reuse later (it can be named `global` or whatever else).

Another option is to create `scripts` folder in your repository that will contain shell scripts and invoke them from the `codemagic.yaml` workflows. 

In both case our goal at Codemagic to make your CI/CD pipeline as flexible as possible and avoid any "vendor-lock", so you can use universal solutions like YAML or shell scripts. I hope it makes sense.

**Reply by mikhail-tokarev** [reply]
just small example to explain `codemagic.yaml` file structure better

```
definitions:
  scripts:
    - &foo
      ...
    - &bar
      ...
  # you can reuse everything not just scripts
  cache: &cache

workflows:
  my-workflow:
    cache: *cache
    scripts:
      - *foo
      - *bar
```

**Reply by Reedyuk** [reply]
i wasnt aware you can use definitions block. I will use this going forward, thanks.

As for the shell scripts, most of my script blocks are literally just calling single gradle tasks. Generally i have put most of my tasks into gradle tasks and then its just a one liner on the command line, but I want to avoid duplicating even that, hence i will use definitions block
